pkgs/api/src/lib.zig

```diff
@@ -58,33 +76,60 @@ pub const Timer = struct {
         // For freestanding targets where we can't measure time, just record 0
         const duration_seconds = if (duration_ns == 0) 0.0 else @as(f32, @floatFromInt(duration_ns)) / 1_000_000_000.0;
 
-        if (self.is_chain) {
-            metrics.chain_onblock_duration_seconds.observe(duration_seconds);
-        } else {
-            metrics.block_processing_duration_seconds.observe(duration_seconds);
-        }
+        self.observeFn(self.context, duration_seconds);
 
         return duration_seconds;
     }
 };
 
 /// A wrapper struct that exposes a `start` function to match the existing API.
 pub const Histogram = struct {
-    is_chain: bool,
+    context: *anyopaque,
+    observeFn: *const fn (*anyopaque, f32) void,
 
     pub fn start(self: *const Histogram) Timer {
         return Timer{
```

Comment:
see if you can make `Timer` a generic type so that you can simply define an `observe` function and call it directly from `Timer.observe`, instead of using a function pointer.